File: src/main/java/nl/kabisa/service/booking/BookingService.java
package nl.kabisa.service.booking;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

import nl.kabisa.service.rating.Rate;
import nl.kabisa.service.rating.RateProvider;
import nl.kabisa.util.Result;

public class BookingService {
    private final List<RateProvider> rateProviders;

    public BookingService(List<RateProvider> rateProviders) {
        this.rateProviders = rateProviders;
    }

    public BookingService(RateProvider... rateProviders) {
        this(Arrays.asList(rateProviders));
    }

    public Result<Booking, BookingException> book(Shipment shipment) {
        return bookShipment(shipment);
    }

    public Booking book2(Shipment shipment) throws BookingException {
        return bookShipment(shipment).unwrap();
    }

    private Result<Booking, BookingException> bookShipment(Shipment shipment) {
        return cheapestRate(shipment)
                .map(rate -> Result.<Booking, BookingException> ok(new Booking(rate)))
                .orElse(Result.err(new BookingException.NoRatesFoundException()));
    }

    private Optional<Rate> cheapestRate(Shipment shipment) {
        return rateProviders.stream()
                .flatMap(provider -> provider.getRates(shipment).stream())
                .min(Rate.cheapest());

    }
}

File: src/main/kotlin/nl/kabisa/service/booking/BookingService.kt
package nl.kabisa.service.booking

import nl.kabisa.service.rating.Rate
import nl.kabisa.service.rating.RateProvider
import java.util.*

class BookingService(private val rateProviders: List<RateProvider>) {

    constructor(vararg rateProviders: RateProvider) : this(Arrays.asList(*rateProviders))

    fun book(shipment: Shipment): BookingResult {
        return bookShipment(shipment)
    }

    @Throws(BookingException::class)
    fun book2(shipment: Shipment): Booking {
        return bookShipment(shipment).unwrap()
    }

    private fun bookShipment(shipment: Shipment): BookingResult {
        val cheapestRate = cheapestRate(shipment)

        return when (cheapestRate) {
            null -> BookingResult.err(BookingException.NoRatesFoundException())
            else -> BookingResult.ok(Booking(cheapestRate))
        }
    }

    private fun cheapestRate(shipment: Shipment): Rate? {
        return rateProviders
                .flatMap({ provider -> provider.getRates(shipment) })
                .minWith(Rate.cheapest())

    }
}
